### PR TITLE
fix: separate Heartbeat and ShardHandler to fibers

### PR DESCRIPTION
**Pull request by kostasrim**
The issue surfaced when the test started failing in #3891. `Heartbeat` will `block indefinitely` if `TriggerJournalWriteToSink` can't `push` the `serialized data` to the channel (if the replica becomes unresponsive). The consequence of that is that the `shard_handler` and the `BreakStalledFlows` will never be called and df will `stall`. The solution is to split the blocking/preemptive flow (Heartbeat) and the `BreakStalledFlows` into separate fibers.

* separate shard_handler from Heartbeat
* add test

Resolves: #3937

### CODE REVIEW

File: src/server/engine_shard.h [hunk 1]
```diff
@@ -207,11 +207,18 @@ class EngineShard {
...
+  void StartPeriodicFiberWithoutHeartbeat(util::ProactorBase* pb,
+                                          std::function<void()> shard_handler);
+  void StartPeriodicFiberImpl(util::ProactorBase* pb, std::function<void()> shard_handler,
+                              bool heartbeat);
 
   void Heartbeat();
   void RetireExpiredAndEvict();
 
-  void RunPeriodic(std::chrono::milliseconds period_ms, std::function<void()> shard_handler);
+  void RunHeartbeatPeriodic(std::chrono::milliseconds period_ms,
```
Comment (adiholden) [inline src/server/engine_shard.h:218]: why does heartbeat still gets shard handler?

File: src/server/engine_shard.cc [hunk 2]
```diff
@@ -391,28 +391,53 @@ void EngineShard::Shutdown() {
...
+  if (heartbeat) {
+    fiber_heartbeat_periodic_ =
+        MakeFiber([this, index = pb->GetPoolIndex(), period_ms = clock_cycle_ms,
+                   handler = std::move(shard_handler)]() mutable {
+          ThisFiber::SetName(absl::StrCat("heartbeat_periodic", index));
+          RunHeartbeatPeriodic(std::chrono::milliseconds(period_ms), std::move(handler));
+        });
+  } else {
+    fiber_shard_handler_periodic_ =
+        MakeFiber([this, index = pb->GetPoolIndex(), period_ms = clock_cycle_ms,
```
Comment (adiholden) [inline src/server/engine_shard.cc:426]: see the condition in origin code RunPeriodic to run the shard handler
if (shard_handler && last_handler_ms + 100 < last_heartbeat_ms)

we run shard_handler every 100 ms this is unrelated to to FLAGS_hz
I believe this function StartPeriodicFiberImpl is rednudant

File: tests/dragonfly/replication_test.py [hunk 3]
```diff
@@ -2317,9 +2319,9 @@ async def test_replication_timeout_on_full_sync(df_factory: DflyInstanceFactory,
 
     await c_replica.execute_command(
         "debug replica pause"
-    )  # puase replica to trigger reconnect on master
+    )  # pause replica to trigger reconnect on master
 
-    await asyncio.sleep(1)
+    await asyncio.sleep(10)
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2324]: is this change needed?

File: tests/dragonfly/replication_test.py [hunk 4]
```diff
@@ -2626,3 +2628,73 @@ async def test_replica_of_replica(df_factory):
...
+    replica = df_factory.create()
+
+    df_factory.start_all([master, replica])
+
+    c_master = master.client()
+    c_replica = replica.client()
+
+    await c_master.execute_command("debug", "populate", "100000", "foo", "5000")
+
+    class ExpirySeeder:
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2651]: put in  utility.py

File: tests/dragonfly/replication_test.py [hunk 5]
```diff
@@ -2626,3 +2628,73 @@ async def test_replica_of_replica(df_factory):
...
+
+        def stop(self):
+            self.stop_flag = True
+
+    c_master = master.client()
+    c_replica = replica.client()
+
+    seeder = ExpirySeeder()
+    seeder_task = asyncio.create_task(seeder.run(c_master))
+    await seeder.wait_until(50000)
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2673]: if the expire seeder will stop setting here wouldnt the bug will reproduce as well?
do we need to continue sending commands to get to the bug?

File: tests/dragonfly/replication_test.py [hunk 6]
```diff
@@ -2626,3 +2628,73 @@ async def test_replica_of_replica(df_factory):
...
+    await c_master.execute_command("debug", "populate", "100000", "foo", "5000")
+
+    class ExpirySeeder:
+        def __init__(self):
+            self.stop_flag = False
+            self.i = 0
+
+        async def run(self, client):
+            while not self.stop_flag:
+                await client.execute_command(f"SET tmp{self.i} bar{self.i} EX 4")
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2658]: use pipeline to speed up things

File: tests/dragonfly/replication_test.py [hunk 7]
```diff
@@ -2626,3 +2628,73 @@ async def test_replica_of_replica(df_factory):
...
+                self.i = self.i + 1
+
+        async def wait_until(self, count):
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2661]: wait_untill_n_inserts or something like this

File: tests/dragonfly/replication_test.py [hunk 8]
```diff
@@ -2626,3 +2628,73 @@ async def test_replica_of_replica(df_factory):
...
+    await c_replica.execute_command(f"REPLICAOF localhost {master.port}")
+
+    # wait for full sync
+    async with async_timeout.timeout(3):
+        await wait_for_replicas_state(c_replica, state="full_sync", timeout=0.05)
+
+    await c_replica.execute_command("debug replica pause")
+
+    # Dragonfly will get stuck here. The journal writes to a channel which will block on write.
+    # Hearbeat() will be called and will get stuck while it tries to evict an expired item (because
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2684]: a. please write "expires" and not "evict" expired item
b. I would not write blocked while pushes to **channel** as this is very specific implemenation detail which also going to change. The reason we get stuck is that we block on write to socket as replica does not reads from it
c. please dont write that BreakStalledFlows() will never be called as this is your bug fix, I would describe the bug and this details in github bug issue

File: tests/dragonfly/replication_test.py [hunk 9]
```diff
@@ -2626,3 +2628,73 @@ async def test_replica_of_replica(df_factory):
...
+
+
+@pytest.mark.asyncio
+async def test_replication_timeout_on_full_sync_heartbeat_expiry(
+    df_factory: DflyInstanceFactory, df_seeder_factory
+):
+    # Timeout set to 3 seconds because we must first saturate the socket such that subsequent
+    # writes block. Otherwise, we will break the flows before Heartbeat actually deadlocks.
+    master = df_factory.create(
+        proactor_threads=2, replication_timeout=3000, vmodule="replica=2,dflycmd=2"
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:2640]: why specify proactor_threads=2 ?

File: tests/dragonfly/replication_test.py [hunk 10]
```diff
@@ -2295,7 +2295,9 @@ async def test_announce_ip_port(df_factory):
 @pytest.mark.asyncio
 async def test_replication_timeout_on_full_sync(df_factory: DflyInstanceFactory, df_seeder_factory):
     # setting replication_timeout to a very small value to force the replica to timeout
-    master = df_factory.create(replication_timeout=100, vmodule="replica=2,dflycmd=2")
+    master = df_factory.create(
+        proactor_threads=2, replication_timeout=100, vmodule="replica=2,dflycmd=2"
```
Comment (kostasrim) [inline tests/dragonfly/replication_test.py:2299]: same here, no need of `proactor_threads=2`

File: tests/dragonfly/replication_test.py (hunk 3 above)
Comment (kostasrim) [reply tests/dragonfly/replication_test.py:2324]: nope, it was accidental

File: src/server/engine_shard.h (hunk 1 above)
Comment (kostasrim) [reply src/server/engine_shard.h:218]: I wanted to keep the callback just in case we want to use it the future but I agree it redundant. I will remove

File: tests/dragonfly/replication_test.py (hunk 3 above)
Comment (kostasrim) [reply tests/dragonfly/replication_test.py:2324]: nope, it was accidental

File: src/server/engine_shard.cc (hunk 2 above)
Comment (kostasrim) [reply src/server/engine_shard.cc:426]: It depends on the condition implicitly when the period is greater than 100 ms because we wait `period_ms`. So if let's say, `period_ms=500` the shard handler will run every `500ms` (we always wait period_ms between runs) so that `100` is really a minimum. I refactored this a little bit to mirror this behaviour

File: tests/dragonfly/replication_test.py (hunk 9 above)
Comment (kostasrim) [reply tests/dragonfly/replication_test.py:2638]: Because it was easier to reproduce with 2 threads. I haven;t tried with more so maybe the test requires tuning.

**Review by adiholden — APPROVED** [review]
